packages/thirdweb/src/react/web/ui/ConnectWallet/screens/Buy/SwapScreen.tsx

```diff
@@ -127,9 +157,51 @@ export function BuyScreenContent(props: {
     });
   };
 
+  const initialAmountNeeded = props.buyForTx
+    ? formatNumber(
+        Number(toEther(props.buyForTx.cost - props.buyForTx.balance)),
+        4,
+      )
+    : 0;
+
   // token amount
-  const [tokenAmount, setTokenAmount] = useState<string>("");
+  const [tokenAmount, setTokenAmount] = useState<string>(
+    initialAmountNeeded === 0 ? "" : String(initialAmountNeeded),
+  );
+
+  // once the user edits the tokenInput or confirms the Buy - stop updating the token amount
+  const [stopUpdatingTokenAmount, setStopUpdatingTokenAmount] = useState(
+    !props.buyForTx,
+  );
+
+  const [amountNeeded, setAmountNeeded] = useState<bigint | undefined>(
+    undefined,
+  );
+
+  // update amount needed every 30 seconds
+  // also update the token amount if allowed
+  // ( Can't use useQuery because tx can't be added to queryKey )
+  useEffect(() => {
+    const buyTx = props.buyForTx;
+    if (!buyTx) {
+      return;
+    }
+    const id = setInterval(() => {
+      getTotalTxCostForBuy(buyTx.tx).then((totalCost) => {
+        setAmountNeeded(totalCost);
```

Comment:
if we're in a new iteration already, do we still want to set the old value when it resolves?

Reply:
done